R/install.R

```diff
@@ -1,276 +1,343 @@
-
-        running <- setdiff(running, installed)
-        bar$message(finished)
-        update_progress(length(installed))
-      }
-      if (!bar$finished && length(running)) {
-        update_progress(0)
-      }
+    while (1) {
+      if (are_we_done(state)) break;
+      ## TODO: update progress bar here
+
+      events <- poll_workers(state)
+      state <- handle_events(state, events)
+      task  <- select_next_task(state)
+      state <- start_task(state, task)
     }
-    events <- get_events(events, num_workers, lib, bar)
+  }, error = function(e) kill_all_processes(state))
+
+  create_install_result(state)
+}
+
+make_start_state <- function(plan, config) {
+
+  ## We store the data about build and installation here
+  install_cols <- data.frame(
+    stringsAsFactors = FALSE,
+    build_done = plan$type == "installed" | plan$binary,
+    build_time = ifelse(plan$binary, 0, NA_real_),
+    build_error = I(rep_list(nrow(plan), list())),
+    build_stdout = I(rep_list(nrow(plan), character())),
+    build_stderr = I(rep_list(nrow(plan), character())),
+    install_done = plan$type == "installed",
+    install_time = ifelse(plan$type == "installed", 0, NA_real_),
+    install_error = I(rep_list(nrow(plan), list())),
+    install_stdout = I(rep_list(nrow(plan), character())),
+    install_stderr = I(rep_list(nrow(plan), character())),
+    worker_id = NA_character_
+  )
+  plan <- cbind(plan, install_cols)
+
+  list(
+    plan = plan,
+    workers = list(),
+    config = config)
+}
+
+are_we_done <- function(state) {
+  all(state$plan$install_done)
+}
+
+#' @importFrom callr poll
+
+poll_workers <- function(state) {
+  if (length(state$workers)) {
+    timeout <- get_timeout(state)
+    procs <- lapply(state$workers, "[[", "process")
+    res <- poll(procs, ms = timeout)
+    vapply(res, function(x) "ready" %in% x, logical(1))
+
+  } else {
+    logical()
   }
-  bar$terminate()
+}
+
+## TODO: No timeout currently
 
-  structure(unlist(events$results, recursive = FALSE), class = "installation_results", elapsed = Sys.time() - start)
+get_timeout <- function(state) -1L
+
+handle_events <- function(state, events) {
+  for (i in which(events)) state <- handle_event(state, i)
+  state$workers <- drop_nulls(state$workers)
+  state
 }
 
-get_events <- function(events, num_workers, lib, bar) {
-  done <- map_lgl(events$processes, function(x) !x$is_alive() && !x$is_incomplete_output() && !x$is_incomplete_error())
-  failed <- map_lgl(events$processes, function(x) !x$is_alive() && x$get_exit_status() != 0)
-  if (any(failed)) {
-    kill_all_processes(events$processes)
-    bar$terminate()
-    events$processes$get_result[which(failed)[[1]]]
+handle_event <- function(state, evidx) {
+  proc <- state$workers[[evidx]]$process
+
+  ## Read out stdout and stderr. If process is done, then read out all
+  if (proc$is_alive()) {
+    state$workers[[evidx]]$stdout <-
+      c(state$workers[[evidx]]$stdout, out <- proc$read_output(n = 10000))
```

Comment:
What happens if there are more than 10000 lines of output? (admittedly unlikely)

Reply:
It reads 10000 lines in this tick of the event loop. The rest is read in the later ticks.